### ISSUE TITLE
Add WorkflowIdConflictPolicy in child workflow options

### ISSUE THREAD

**Issue by Quinn-With-Two-Ns** [temporalio/features#558]
### Is your feature request related to a problem? Please describe.

Add `WorkflowIdConflictPolicy` in child workflow options

### Additional context

https://github.com/temporalio/features/issues/437

### Per-SDK Tickets

- [ ] Go - 
- [ ] Java - 
- [ ] Core - 
- [ ] TypeScript - 
- [ ] Python - 
- [ ] .NET - 
- [ ] Ruby - 
- [ ] PHP - 
- [ ] Temporal CLI -

**Comment by madhav2302** [comment]
I agree with @lukeramsden
I am not convinced with the argument provided as well. 

ChildWorkflows are more of a reference for our use-case, rather than real parent-child relation. 

It is easier to build DAG based workload if we can simply re-attach existing running workflow, rather than going through the ugly workaround of signals, and running child workflow from an activity.

**Comment by cretz** [comment]
> USE_EXISTING from inside a workflow does not necessarily have to re-parent the targeted workflow, surely?

It does, because the lifetimes (e.g. parent close policy) and cancellation events concerning child workflows are very specific to child workflows and are expected to apply to whatever start/execute child workflow returns. Child workflows are bound to their parents. I suppose we could say "if you use 'abandon' parent close policy and promise to never need to cancel the child workflow, then you could use use-existing conflict policy" but at that point, why are you using child workflows instead of top-level workflows started via a Temporal client in an activity? If you do want to react to a "workflow already exists" conflict for children in whatever way, all SDKs already provide a very specific error for this that is returned when you try to start a child. The use-existing conflict policy just becomes sugar that doesn't even support all child features.

**Comment by madhav2302** [comment]
> I suppose we could say "if you use 'abandon' parent close policy and promise to never need to cancel the child workflow, then you could use use-existing conflict policy"

Yes, it can be mandatory if using child workflows with workflow id conflict policy for child workflows. 

> at that point, why are you using child workflows instead of top-level workflows started via a Temporal client in an activity?

The idea is to have cleaner interface, and readability of flow instead of using activity and signals as workaround, which is fine if you use it for very small use-case, but it makes very sense to have it for major business use-case. 

> If you do want to react to a "workflow already exists" conflict for children in whatever way, all SDKs already provide a very specific error for this that is returned when you try to start a child.

It's not being reactive, but usually DAG workload needs to wait until the child completes. 

Of-course, this can be solved if you use `updateWithStart`, but still it needs to be supported via workflow code.

**Comment by cretz** [comment]
That is a separate issue at https://github.com/temporalio/temporal/issues/680

**Comment by madhav2302** [comment]
@cretz Ideally, we would want an atomic way of starting (if not exist) and waiting for the result. ChildWorkflow with WorkflowIdConflictPolicy sounded like the idea, but we have no issues if we can simply start another workflow with WorkflowIdConflictPolicy option from workflow code and we don't call it child workflow.

**Comment by prasek** [comment]
more info on ☝️ https://docs.temporal.io/nexus/operations#attaching-multiple-nexus-callers

**Comment by prasek** [comment]
> We are using java, are there [Nexus] examples of it we can refer to?

[Here’s the tutorial](https://docs.temporal.io/develop/java/nexus) and [sample](https://github.com/temporalio/samples-java/tree/main/core/src/main/java/io/temporal/samples/nexus). The conflict policy belongs on the `WorkflowOptions` in the [Nexus Service implementation](https://github.com/temporalio/samples-java/blob/main/core/src/main/java/io/temporal/samples/nexus/handler/NexusServiceImpl.java#L50)

```
.setWorkflowIdConflictPolicy(
    WorkflowIdConflictPolicy.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING)
```

as mentioned in [the docs](https://docs.temporal.io/nexus/operations#attaching-multiple-nexus-callers):
> Nexus Operations that start a Workflow with the [New-Workflow-Run-Operation](https://docs.temporal.io/nexus/operations#sdk-support) SDK helper will automatically attach a completion Callback on the handler Workflow, so the Nexus caller receives the result. Additional Nexus callers may attach to the same handler Workflow if the Nexus handler uses a [Conflict-Policy of Use-Existing](https://docs.temporal.io/workflow-execution/workflowid-runid#workflow-id-conflict-policy).

> A single handler Workflow Execution has a [Workflow Callback limit](https://docs.temporal.io/workflow-execution/limits#workflow-execution-callback-limits) that governs how many Nexus callers can be attached. Nexus callers that exceed the limit will receive an error.

Note this this is not a Nexus limitation, but the Workflow Callback limit (currently 32) affects how many Nexus callers can attach to the same underlying handler Workflow Execution. Workflow Callbacks support state-based delivery with auto-retries, circuit breaking, rate limiting and so on.

In the near future we're looking to support > 32 Nexus callers with manual Nexus completion callback delivery. This will allow you to opt-out of the built-in Workflow Callback mechanism by manually storing the Nexus callback tokens in a handler Workflow variable (or external storage), and then manually delivering Nexus completion callbacks via a new Temporal SDK `CompleteNexusOperation` method that should be called just prior to handler Workflow completion. Note this approach will scale to a very large number of Nexus callers.

Longer term we're looking to significantly raise the built-in [Workflow Callback limit](https://docs.temporal.io/workflow-execution/limits#workflow-execution-callback-limits), which should cover all but the largest scenarios and provide the best ergonomics.

In short:
1. there's a limit of 32 Nexus callers that can attach to a single handler Workflow Execution today
2. we're working on supporting > 32 Nexus callers with manual Nexus completion callback delivery
3. longer term we're looking into substantially increasing the number of built-in Workflow Callbacks supported

**Comment by madhav2302** [comment]
> This will allow you to opt-out of the built-in Workflow Callback mechanism by manually storing the Nexus callback tokens in a handler Workflow variable (or external storage), and then manually delivering Nexus completion callbacks via a new Temporal SDK CompleteNexusOperation method that should be called just prior to handler Workflow completion. Note this approach will scale to a very large number of Nexus callers.

Does this mean it's a manual signal way implementation where the user needs to maintain it? It sounds very similar to me as a signal.

It will defeat the purpose of atomic completion of caller workflow , as what if the called workflow terminated due to any reason, and now it's not able to send the completion signal back. The caller workflow will keep waiting indefinitely. 

Also, the opt-out is also not nice due to 99% of our workload can be satisfies the limits and we have no way to determine which workload needs to be opt in or opt out when they start. 

It would be nice if we you can convert it into automatic manual handler when we hit the limit, and if that's not implemented, simply throw an error. 

> longer term we're looking into substantially increasing the number of built-in Workflow Callbacks supported
Would be nice to put some timelines on it, as it will help us plan our efforts accordingly.
